### PR TITLE
Paper: VisPy: Harnessing The GPU For Fast, High-Level Visualization (Campagnola)

### PR DESCRIPTION
**Pull request by campagnola**
Conference proceedings for VisPy.

I had some trouble with table and reference formatting--the wide table needs to be wider (and writing the table directly as latex will make it difficult to get syntax highlighting inside the table). Also tried to add proper labels / references for figures and tables, but these prevented the document from compiling. I hope we can sort those out soon.

Finally, I will be adding 2-3 more figures, but these are contingent upon finishing some new features in VisPy. If the features are not ready soon, then I will just add what we have available.

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/3

### CODE REVIEW

File: papers/luke_campagnola-vispy/vispy.rst [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+.. class:: keywords
+
+   graphics, visualization, plotting, performance, interactive, opengl 
+
+
+Motivation
+----------
+
+Despite the rapid growth of the scientific Python stack, one conspicuously absent element is a standard package for high-performance visualization. The de-facto standard for plotting is Matplotlib; however, this package is designed for publication graphics and is not optimized for visualizations that require realtime, interactive performance, or that incorporate large data volumes. Several packages have apeared in the Python ecosystem to fill this gap (vtk, chaco, pyqtgraph, visvis, galry, glumpy, etc.), but none has yet emerged as a serious candidate standard package. Consequently, these projects have spent much of their effort re-solving the same problems, and the ecosystem as a whole suffers from the lack of a focused, collaborative effort.
+
```
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: It is inaccurate and incorrect to put VTK in the same list as chaco, pyqtgraph, or any of the other packages.  VTK provides way more than just graphics.  I agree it is hard to build and distribute but there are cross platform builds of the package available relatively easily today.  If anything, by trying to redo what VTK does in Python it will be other libraries that are re-solving the same problems.

Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: I'm not sure I understand why it is relevant that VTK provides more than graphics. PyQtGraph also provides more than graphics, but the point here was just to list the packages that are being used by the Python community for fast graphics. 

I should also point out that we are not trying to redo VTK, which uses the old fixed-function OpenGL pipeline (and when I last spoke to the kitware folks, there were no plans to upgrade). By making use of the modern shader pipeline, vispy will be able to put much more of the computing effort on the GPU, and offer better quality at the same time.

File: papers/luke_campagnola-vispy/vispy.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+What is VisPy
+-------------
+
+VisPy is a scientific visualization library based on OpenGL. Its primary purpose is to deliver high-performance rendering under heavy load, but at the same time we aim to provide publication-quality graphics, a high-level 2D and 3D plotting API, and portability across many platforms. VisPy's main design criteria are:
+    
+* *High-performance for large data sets.* By making use of the modern, shader-based OpenGL pipeline, most of the graphical rendering cost is offloaded to the graphics processor (GPU). This allows realtime interactivity even for data on the order of millions of samples, and at the same time minimizes CPU overhead.
+  
+* *High-level visualization tools.* Most Python developers are not graphics experts. Getting from raw data to interactive visualization should require as little code as possible, and should require no knowledge of OpenGL or the underlying graphics hardware.
+  
+* *Publication quality output.* Commodity graphics hardware and the modern OpenGL shader pipeline have made it possible to render moderately large data sets without sacrificing quality.
```
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: This is not entirely the same as publication quality which often requires vector graphics for all or part of the graphics being rendered.  I am not sure this has anything to do with the shader pipeline.

File: papers/luke_campagnola-vispy/vispy.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+--------------------
+
+VisPy's functionality is divided into a layered architecture, with each new layer providing higher-level primitives. The top layers provide a powerful system for quickly and easily visualizing data, whereas the lower layers provide greater flexibility and control over OpenGL's features.
+
+
+Layer 1: Object-Oriented GL
+'''''''''''''''''''''''''''
+
+The OpenGL API, although very powerful, is also somewhat verbose and unwieldy. VisPy's lowest-level layer, ``vispy.gloo``, provides an object-oriented OpenGL wrapper with a clean, compact, and Pythonic alternative to traditional OpenGL programming (however, developers unfamiliar with OpenGL are encouraged to work from the scenegraph and plotting layers instead). Objects that typically require several GL calls to instantiate, such as textures, vertex buffers, frame buffers, and shader programs, are instead encapsulated in simple Python classes (e.g. Table 1).
+
```
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: Unfortunately, the table does not render cleanly in the PDF, it would be great if you could please check this.  Thanks!

File: papers/luke_campagnola-vispy/vispy.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+.. figure:: image_transforms.png
+
+   One image viewed using four different coordinate transformations. VisPy supports linear transformations such as scaling, translation, and affine matrix multiplication (bottom left) as well as nonlinear transformations such as logarithmic (top left) and polar (top right). Custom transform classes are also easy to construct (bottom right).
+
+The following example summarizes the code that produces the logarithmically-scaled image in Figure XX. It combines a scale/translation, followed by log base 2 along the y axis, followed by a second scale/translation to set the final position on screen. The resulting chained transformation maps from image coordinates (origin in upper left, 1 unit = 1 image pixel) to window coordinates (origin in upper left, 1 unit = 1 window pixel):
```
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: Could you please correct the figure reference?

File: papers/luke_campagnola-vispy/vispy.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+   
+   # Plot data in the first grid cell
+   fig[0, 0].plot(data)
+   
+   # Add a spectrogram of the same data in the next row
+   fig[1, 0].spectrogram(data)
+
+Despite the large volume of data, the resulting views can be immediately panned and zoomed in realtime. As a rough performance comparison, the same plot data can be redrawn at about 0.2 Hz by Matplotlib, 2 Hz by PyQtGraph, and over 100 Hz by VisPy (on the author's machine). 
+
+Each function in ``vispy.ploy`` generates scenegraph (layer 3) objects to allow lower level control over the visual output. This makes it possible to begin development with the simplest ``vispy.plot`` calls and iteratively refine the output as needed. VisPy also includes an experimental wrapper around ``mplexporter`` (from https://github.com/mpld3/mplexporter) that allows it to act as a drop-in replacement for Matplotlib in existing projects (however this approach is not always expected to have the same performance benefits as using the native ``vispy.plot`` API).
```
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: Do you mean ``vispy.plot`` instead of ``vispy.ploy`` in the first line?

File: papers/luke_campagnola-vispy/vispy.rst (hunk 1 above)
Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: PyQtGraph does not really concern itself with "visualization algorithms" in 3D.  If you want to read a variety of datasets, contour them, slice through them, extract subsets of them, handle multi-grid data, handle structured and unstructured grids, that is all in the domain of VTK, and not part of the other libraries.  VTK is a very different beast from galry or glumpy or visvis, its scope is much broader.  VTK 6.2 and beyond will be using the programmable pipeline (it is still experimental but available if you know what to build and your hardware supports it).

Comment (prabhuramachandran) [inline papers/luke_campagnola-vispy/vispy.rst]: Please do not misunderstand my comment, I am not arguing against vispy.  I really like the project and I think vispy is excellent.  I just strongly disagree with that statement and do not think VTK and the others are comparable in sense of the sentence.

**Comment by prabhuramachandran** [comment]
vispy is a very interesting project and a much needed effort as mentioned in the paper.  The project holds a lot of promise and I am excited to see it and get my hands dirty -- I've checked out the project a few times in the past. In particular it is really heartening to see a collaborative effort on this. I think the paper can bring out the nice features of vispy in a much better way.  More non-trivial high-level scripts that show the following would be really nice:

 1. Large data which is not just a scatter of points.
 2. An example that shows how easy things work on a GUI widget or IPython notebook or for an off-screen context.  This fact should be emphasized.
 3. An example that brings out the ease of interaction and events that can be written in pure Python (picking, or clicking if implemented).
 4. Any other features that would interest the SciPy user/reader/developer.

Similarly, 
 1. A more detailed exposition on the higher level features planned would be nice to showcase.  Mayavi's `mlab` already has a set of simple (and limited) features that can compared with.
 2. More details on how the vagaries of supported hardware and the abstraction layers handle those would be nice.  Some details of how the backends abstract out the differences between OpenGL and WebGL would be good to see.

One important issue in the past I have seen is that different hardware support different GLSL versions (and vispy did not handle them transparently), how do you handle that without repeating code or rewriting the basic pieces?  It would be nice if these architectural pieces were discussed.  How will older hardware without a GPU work?  Will it be seamless or will it not work at all?  These are important considerations that should be discussed. I think it is important to show the strength of the architecture in a paper like this as it will help any future contributors.

In summary, the paper did not make me sit up and take notice and I really ought to given the laudable goals of the vispy project.  Thank you.

**Comment by dalippa** [comment]
@campagnola I was asked to join as an additional reviewer. I've read the paper and have some feedback to offer.

1. There are a few claims you make that could probably use a cited source. For example:
a. "Several packages have appeared... but not has yet emerged as a serious cadide standard package"
b. "these projects have spent much of their effort re-solving the same problems..."

2. "VisPy strives to make common tasks easy"... What are the common tasks? Can you elaborate?

3. I am intrigued by the portability claim: I would like to see more details on how you did this in pure python with only a dependence on numpy!

4. Table 1 has formatting issues. See http://nipy.bic.berkeley.edu:5000/download/3

5. There are a number of places where you break the flow of the paper using parentheses, which impedes readability. Please consider footnotes as an alternative.

6. How do you handle issues involving aspects of OpenGL that use global state (e.g. glTexEnv)

7. There are some domain-specific abbreviations (such as GLSL) that are used before they are defined.

8. Your citation of Table 1 should probably be on page 2, and Figure 1, cited later on that page, should probably be after it or on the next page. Figure descriptions could use to be a bit more concise.

9. Some sentences in the "Layer 2" section are quite long and might be better broken up.

10. Is the code snippet on page 2 complete? Page 3 starts with Table 1, which looks like it's continuing from the example on the previous page.

11. Consider changing Roadmap to "future work". The roadmap section seems to drop off quickly and not transition well.

12. Paper is completely missing any citations, and could benefit from a citation containing the github repository and/or community pages.

**Comment by campagnola** [comment]
Thanks very much @prabhuramachandran and @dalippa for the excellent and thoughtful reviews!
Edits and new figures forthcoming.

File: papers/luke_campagnola-vispy/vispy.rst (hunk 1 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: I've adjusted the wording to reflect the differences between the libraries and also to sound a bit less competitive. Please let me know if this doesn't address your concern, though..

File: papers/luke_campagnola-vispy/vispy.rst (hunk 2 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: Added a note on the possibility of vector graphics output.

File: papers/luke_campagnola-vispy/vispy.rst (hunk 1 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: I guess GitHub is not going to make this easy.. here's the relevant change:

> Several packages exist in the Python ecosystem that are capable of high-performance visualization: VTK provides an extensive set of 3D visualization tools with Python bindings, Chaco  offers efficient 2D plotting, PyQtGraph is a scientific GUI library with fast ploting, Glumpy implements high-quality plotting primitives in OpenGL, and VisVis and Galry both provide high-performance 2D/3D OpenGL visualization. Each of these packages has its particular strengths and weaknesses and, although the Python community benefits from such a rich ecosystem, at the same time it suffers from the lack of a focused, collaborative effort.

File: papers/luke_campagnola-vispy/vispy.rst (hunk 2 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: > *Publication quality output.* Commodity graphics hardware and the modern OpenGL shader pipeline have made it possible to render moderately large data sets without sacrificing quality in primitive shapes or antialiasing. VisPy is also designed to enable vector graphics output, although this feature is not yet implemented.

File: papers/luke_campagnola-vispy/vispy.rst (hunk 3 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: I haven't figured out yet how to get the table to full width. The examples suggest just rewriting the table in latex, but that will make the embedded syntax highlighting difficult. Might consider dropping the table if it's too much work..

File: papers/luke_campagnola-vispy/vispy.rst (hunk 5 above)
Comment (campagnola) [inline papers/luke_campagnola-vispy/vispy.rst]: Thanks, good catch.

**Comment by campagnola** [comment]
@prabhuramachandran:

> Large data which is not just a scatter of points.

    Added a scrolling plots example in the optimization section; I hope this fits.

> An example that shows how easy things work on a GUI widget or IPython notebook or for an off-screen context. This fact should be emphasized.

    No example, but I did address this in the `gloo` section.

> An example that brings out the ease of interaction and events that can be written in pure Python (picking, or clicking if implemented).

    I have added a picking example and figure.

> A more detailed exposition on the higher level features planned would be nice to showcase. Mayavi's mlab already has a set of simple (and limited) features that can compared with.

    Future work has been expanded.

> More details on how the vagaries of supported hardware and the abstraction layers handle those would be nice.
> Some details of how the backends abstract out the differences between OpenGL and WebGL would be good to see.
> One important issue in the past I have seen is that different hardware support different GLSL versions (and vispy did not handle them transparently), how do you handle that without repeating code or rewriting the basic pieces?
> How will older hardware without a GPU work? 
> Will it be seamless or will it not work at all?

    These have been addressed in the `gloo` section and more extensively in the new `quality / optimization` section.

**Comment by campagnola** [comment]
@dalippa:

There are a few claims you make that could probably use a cited source. For example:
> a. "Several packages have appeared... but not has yet emerged as a serious cadide standard package"
> b. "these projects have spent much of their effort re-solving the same problems..."

I have removed these claims and softened the intro a bit.

> "VisPy strives to make common tasks easy"... What are the common tasks? Can you elaborate?

Updated to read "VisPy strives to make common tasks easy--most basic plot types can be generated with just a few lines of code."

> I am intrigued by the portability claim: I would like to see more details on how you did this in pure python with only a dependence on numpy!

The dependency claim is "depends only on NumPy and a suitable GUI library". Because we can hand array data directly to OpenGL, there is little need for heavily optimized python routines, and thus little need for external dependencies. This is described a bit more in the Visuals section.

> Table 1 has formatting issues. See http://nipy.bic.berkeley.edu:5000/download/3

I am so far unable to resolve this. Probably the table will just be removed; it is not critical.

> There are a number of places where you break the flow of the paper using parentheses, which impedes readability. Please consider footnotes as an alternative.

One better: I have removed the parentheticals entirely (or merged them into the text sans parentheses).

> How do you handle issues involving aspects of OpenGL that use global state (e.g. glTexEnv)

Added a note: "Before drawing, each visual also configures the necessary OpenGL global state such as blending and depth testing. These state parameters may be reconfigured for each visual to select different compositing modes."

> There are some domain-specific abbreviations (such as GLSL) that are used before they are defined.

I think these are now all fixed.

> Figure descriptions could use to be a bit more concise.

> Some sentences in the "Layer 2" section are quite long and might be better broken up.

I have broken up some longer sentences and rewritten some if this section. Please let me know if there are specific sentences that still stand out to you.


> Is the code snippet on page 2 complete? Page 3 starts with Table 1, which looks like it's continuing from the example on the previous page.

> Consider changing Roadmap to "future work". The roadmap section seems to drop off quickly and not transition well.

Done, expanded, and concluded.

> Paper is completely missing any citations, and could benefit from a citation containing the github repository and/or community pages.

Done.

**Comment by campagnola** [comment]
The only remaining issue I see is the table. It requires only a minor modification to correct the column widths in the tex file:

    \begin{longtable*}[c]{p{0.393\DUtablewidth}p{0.547\DUtablewidth}}

 Anyone know if the .tex can be post-processed? I have tried taking the LaTeX output of the table and copying it directly into a `.. raw:: latex` block, but then there seem to be some pieces missing from the preamble as well. 

@katyhuff  @jaberg

**Comment by dalippa** [comment]
@campagnola A few minor comments:

* Intro looks good now. One sentence,"Several packages exist in the Python..." is a bit unwieldy.

* Table 1 is still formatted weirdly. Perhaps @katyhuff has a suggestion? Removing is fine as well.

* Removal of many of the parentheticals has been completed well. A few remain, but they do not impede readability.

* I only found one more sentence in the Layer 2 section that still stands out as a little long/unwieldy: "VisPy’s  approach  is  for  each  visual  to  implement  multiple...".

* mplexporter citation should be consistent with the others.

* "teselation" was misspelled.

One last note: The paper is well written and conveys excitement. I fully intend on testing out VisPy in the near future.

**Comment by campagnola** [comment]
Thanks @dalippa! I slightly de-awkwarfied the long intro sentence, but ultimately, breaking it into several sentences did not help much. I'll consider suggestions if you have them, but otherwise leave it as is. All other comments have been addressed and I made a few other spelling corrections as well.

**Comment by dalippa** [comment]
@campagnola I just finished re-reviewing. Aside from making the list semicolon delimited - with a marginal gain of readability - there isn't anything that jumps out at me. I recommend leaving as is.

**Comment by dalippa** [comment]
@katyhuff @jaberg After re-reviewing, I recommend acceptance of the paper into the proceedings.

**Comment by campagnola** [comment]
@katyhuff @jaberg, do you think we will be able to correct the table width, or should I remove it?

**Comment by katyhuff** [comment]
@campagnola I think rst isn't really equipped to handle wide code within a table. Perhaps you can persent the code inline with the text, presenting the shortened version followed by the lengthy version, or vice versa. Another option would be to present the comparison as an image (of a two pane terminal, for instance).

So, get rid of the table. I personally tried to debug it for the last 20 minutes and it's just not going to work. Sorry about that.

**Comment by dalippa** [comment]
What about an appendix at the end of the document? Appendix A: the old way.
appendix B: VisPy?

It isn't side by side but it would keep the content there...

On Monday, July 6, 2015, Katy Huff <notifications@github.com> wrote:

> @campagnola <https://github.com/campagnola> I think rst isn't really
> equipped to handle wide code within a table. Perhaps you can persent the
> code inline with the text, presenting the shortened version followed by the
> lengthy version, or vice versa. Another option would be to present the
> comparison as an image (of a two pane terminal, for instance).
>
> So, get rid of the table. I personally tried to debug it for the last 20
> minutes and it's just not going to work. Sorry about that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/159#issuecomment-119058777>
> .
>


-- 
David Lippa
Principal Software Engineer
david.lippa@realmassive.com
RealMassive <http://www.realmassive.com/>
t: @realmassive


* Commercial Real Estate in Real-Time*

*-----*
This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system. E-mail transmission cannot be
guaranteed to be secure or error-free as information could be intercepted,
corrupted, lost, destroyed, arrive late or incomplete, or contain viruses.
The sender therefore does not accept liability for any errors or omissions
in the contents of this message, which arise as a result of e-mail
transmission. If verification is required please request a hard-copy
version.

**Comment by campagnola** [comment]
Thanks @katyhuff, I have inlined the code example. I think this is ready to go.

**Comment by campagnola** [comment]
Thanks Katy!
